DEVOPS-693: migrate from conda-lock to pixi

saicheranb opened the pull request:
**DEVOPS-693 - migrate from conda-lock to pixi**

saicheranb committed:
DEVOPS-693:Refactor pyproject.toml to integrate Pixi tool

- Removed previous dev dependencies under [tool.poetry.group.dev.dependencies].
- Added new [tool.pixi.project] section with channels and platforms.
- Defined direct and indirect dependencies under [tool.pixi.dependencies].
- Specified PyPI options and dependencies for Pixi under [tool.pixi.pypi-options] and [tool.pixi.pypi-dependencies].
- Created environments for default and production under [tool.pixi.environments].
- Added build and test features with respective dependencies and tasks.

saicheranb committed:
Update Python analysis workflow to use Pixi package manager and specific version for static analysis and pytest

sebhmg commented on .codespellignore:
```diff
@@ -1,2 +1,3 @@
 ba
 nd
+astroid
```
**issue:** instead, add exclusion for pixi.lock in the spellcheck task of pre-commit

sebhmg commented on .gitattributes:
```diff
@@ -2,3 +2,5 @@
 *.h5 filter=lfs diff=lfs merge=lfs -text
 *.min.js filter=lfs diff=lfs merge=lfs -text
+# SCM syntax highlighting & preventing 3-way merges
+pixi.lock merge=binary linguist-language=YAML linguist-generated=true
```
**suggestion::** not a binary file


```suggestion
pixi.lock linguist-language=YAML linguist-generated=true
```

sebhmg replied:
**suggetion:** also apply the linguist attributes to poetry and conda lock files

👍 I did not know about github linguist

sebhmg commented on pixi.lock:
**suggestion:** with pixi lock, no more need for conda lock files or `environments` folder

remove all conda lock files and `environments` folder

sebhmg replied:
so far, it was forced by py-deps-lock by setting environment variables:
```sh
            CONDA_CHANNEL_ALIAS="https://repo.prefix.dev"
            MAMBA_CHANNEL_ALIAS="https://repo.prefix.dev"
```

sebhmg replied:
maybe simply set eplicitly the url to prefix.dev for conda-forge:
```yaml
[workspace]
channels = [https://repo.prefix.dev/conda-forge]
```

sebhmg commented on pixi.lock:
```diff
@@ -0,0 +1,2906 @@
+environments:
+  default:
+    channels:
+    - url: https://conda.anaconda.org/conda-forge/
```
**issue:** configure pixi to use prefix.dev channels rather than anaconda.org

sebhmg commented on .gitignore:
```diff
@@ -147,3 +147,6 @@ pyproject-sha.toml
 geoapps_utils/_version.py
+# pixi environments
+.pixi/*
+!.pixi/config.toml
```
**issue**: if `.pixi/config.toml` is to be preserved, it seems missing from this PR

sebhmg commented on pyproject.toml:
```diff
@@ -130,6 +110,51 @@ initial-content = """
+
+[tool.pixi.pypi-dependencies]
+# PyPI / git-based dependencies
+geoh5py = { git = "https://github.com/MiraGeoscience/geoh5py.git", rev = "develop" }
```
**thought:** ideally, with pixi, we would define several environments which would resolve dependencies differently geoh5py:
- from pypi
- from jfrog
- from github
- from relative path on disk

See https://pixi.sh/latest/tutorials/multi_environment/#lets-get-started

sebhmg commented on pyproject.toml:
```diff
@@ -130,6 +110,51 @@ initial-content = """
+pypi-dependencies = { hatch = "*", build = "*", twine = "*" }
+
+[tool.pixi.feature.test.pypi-dependencies]
+Pygments = "*"
+pylint = "*"
+pytest = "*"
+pytest-cov = "*"
+pyyaml = "*"
+jinja2 = "*"
+packaging = "*"
+tomli = "*"
```
**suggestion:** get these test dependencies though conda, rather than through pip

Since we want the final production environment to be conda, it makes more sense getting as much as possible from conda (and that what how it used to be with conda-lock)

sebhmg commented on pyproject.toml:
```diff
@@ -130,6 +110,51 @@ initial-content = """
+jinja2 = "*"
+packaging = "*"
+tomli = "*"
+geoapps-utils = { path = ".", editable = true }
```
**question:** is this required? is not pylint already adding current folder to PYTHON_PATH?

sebhmg reviewed (changes requested):
please, see some questions and suggestions

saicheranb committed:
Refactor dependencies in pyproject.toml

- Updated geoh5py dependency to use the latest version from PyPI instead of a specific git revision.
- Renamed section for test dependencies for clarity.
- Removed geoapps-utils dependency from editable mode.

sebhmg reviewed (changes requested):
please, see comment about declaring dependencies

saicheranb committed:
Update pyproject.toml: Refactor dependencies and change conda channel

- Removed specific version constraints for numpy and scipy.
- Added pydantic as a dependency with version constraints.
- Updated conda channel from "conda-forge" to "https://repo.prefix.dev/conda-forge".

saicheranb committed:
Refactor pyproject.toml: Update dependencies and reorganize project URLs

sebhmg reviewed (changes requested):
see comments about dependencies

sebhmg commented on pyproject.toml:
```diff
@@ -61,11 +61,9 @@ include = [
-pydantic = "^2.5.2"
-scipy = "~1.14.0"
+dependencies = [
+  "pydantic>=2.5.2,<3.0",
```
when specifying geoh5py dependency on pip (not git), we used to have the option --allow-prerelease with Poetry.
This option was really to compute the lock. I could not find an equivalent for pixi.
I assume it is OK without it.

saicheranb committed:
Update dependencies and refine environment configurations in pyproject.toml

- Added geoh5py dependency with version constraints.
- Updated numpy and scipy dependencies.
- Removed commented-out pip dependencies for clarity.
- Enhanced environment configurations to include dev-local and dev-git setups.
- Specified geoh5py source for dev-local and dev-git environments.

saicheranb committed:
Update dependencies in pyproject.toml: Upgrade pydantic version and adjust environment configurations

saicheranb committed:
Update workflow references in python_analysis.yml to use latest reusable workflows # work around

saicheranb committed:
Update workflow references in python_analysis.yml to use DEVOPS-584 reusable workflows

sebhmg commented on pyproject.toml:
```diff
@@ -127,13 +116,20 @@ h5py = ">=3.2.1, <4.0.dev"      # from geoh5py
-
-[tool.pixi.pypi-dependencies]
-geoh5py = "*"
+index-strategy = "unsafe-best-match"
```
**suggestion:** do use JFrog yet here on this repo

I think we can reserve puling from JFrog for our global environment. Or maybe for private repos only.
But on public repo, non-mira developers have no access and are unable to use it


```suggestion
```

sebhmg reviewed (dismissed):
please see comment: do not have options (yet) for Artifactory

saicheranb committed:
Update pyproject.toml

Co-authored-by: Sébastien Hensgen <sebastienh@mirageoscience.com>

saicheranb committed:
Update pyproject.toml

saicheranb committed:
Remove 'dev-local' environment from pixi configurations in pyproject.toml

sebhmg committed:
[DEVOPS-693] adjust pixi environments

- comment out the production environment for now (future versions not published yet)
- commented placeholder for a production+tests environment
- no need for the build feature (yet?)
- "default" env being a copy of "dev-git", make sure dependencies are resolved the same way
- restore the definition of "dev-local" environment

sebhmg committed:
[DEVOPS-693] Revert "Remove 'dev-local' environment from pixi configurations in pyproject.toml"

This reverts commit 8a8dfdc083feaf67b415dd0696d48cbf7618be20.

sebhmg committed:
[DEVOPS-693] relocked with newer pixi (v1.54.1)

sebhmg committed:
Merge pull request #155 from sebhmg/DEVOPS-693

DEVOPS-693: migrate from conda-lock to pixi

sebhmg committed:
[DEVOPS-693] installing with --lock and dev-local fails

as dev-local assumes ../geoh5py exists, it fails with "locked" option on CI.
Thus remove dev-local, and use a task instead for force use of local dependencies

sebhmg committed:
[DEVOPS-693] remove dev-git environment: default is enough

sebhmg committed:
[DEVOPS-693] (cosmetics) remove extra empty lines

sebhmg committed:
[DEVOPS-693] get pydantic from conda, no pypi

sebhmg committed:
[DEVOPS-693] try dependency-overrides for local path

patch with task `use-local-deps` is lost upon every pixi command,
such as pixi run test

sebhmg committed:
[DEVOPS-693] with dependency-overrides, pixi always things the lock is not up-to-date

sebhmg committed:
[DEVOPS-693] add a specific task to run test with local deps

sebhmg committed:
[DEVOPS-693] define a composite task to run test with local deps

sebhmg committed:
[DEVOPS-693] accept space separated list of packages for tasks

sebhmg committed:
[DEVOPS-693] aesthetics: move test task first

sebhmg reviewed (changes requested):
pixi is running all tests with Python 3.12, no matter what is the version used in python-setup

<img width="987" height="282" alt="Image" src="https://github.com/user-attachments/assets/5f080866-e1b7-4051-b7fb-f99df0a84d91" />

sebhmg committed:
[DEVOPS-693] define one environment per python version

sebhmg reviewed (dismissed):
now, it seems it always runs in the default py310 environment

<img width="1278" height="192" alt="Image" src="https://github.com/user-attachments/assets/349c738c-5bf1-4499-b298-1482e98d2257" />

sebhmg committed:
[DEVOPS-693] aesthetics

sebhmg committed:
[DEVOPS-693] use again github workflows from CI-tools @main

sebhmg committed:
[DEVOPS-693] simplify version specifiers

and fix mismatch for pydantic
(also move it under comment block for direct dependencies)

sebhmg committed:
[DEVOPS-693] no need for jinja template in tests

sebhmg committed:
[DEVOPS-693] split and trim dev dependencies

sebhmg commented on tests/version_test.py:
```diff
@@ -15,23 +15,16 @@
-    rendered_yaml = template.render()
-
-    recipe = yaml.safe_load(rendered_yaml)
+def get_conda_recipe_version() -> str | None:
```
no rename  here: it is still a conda recipe

sebhmg committed:
[DEVOPS-693] add package itself to the pixi envs

sebhmg committed:
[DEVOPS-693] always include self

To work with the published package, users would simply run conda install.
Usage of pixi here is limited to working with the local source of the package

sebhmg committed:
[DEVOPS-693] no need for a self feature

just add local package to pixi.pypi-deps

sebhmg commented on pyproject.toml:
```diff
@@ -130,6 +97,86 @@ initial-content = """
+geoh5py = { git = "https://github.com/MiraGeoscience/geoh5py.git", rev = "develop" }
+
+#[tool.pixi.feature.mirageo.dependencies]
+#geoh5py = ">=0.12.0a1, 0.12.*"
```
nope, that a corret syntaxe for conda

sebhmg committed:
[DEVOPS-693] add a task for a repl shell (ipython)

sebhmg committed:
[DEVOPS-693] re-apply recommendations by pixi for lock file

sebhmg committed:
[DEVOPS-693] allow commit of .pixi/config.toml if any

sebhmg committed:
[DEVOPS-693] remove irrelevant entries from gitattribute

sebhmg committed:
[DEVOPS-693] group features and add test-pyvers tasks

sebhmg committed:
[DEVOPS-693] add geoh5 in keywords

sebhmg committed:
[DEVOPS-693] fix indirect dependency versions

As indirect dependencies come through pip, and we force them from conda,
their versions must be specified, or they will conflict with the most recent versions from conda

sebhmg committed:
[DEVOPS-693] add earth-science as keyword in package meta info

sebhmg committed:
[DEVOPS-693] move pixi config to its own file

less risk to break python.toml because error in large config

also move ipython to dev dependencies

sebhmg committed:
[DEVOPS-693] minor: shorter comment

sebhmg committed:
[DEVOPS-693] aesthetics

sebhmg committed:
[DEVOPS-693] add keywords and classifiers troves

sebhmg committed:
[DEVOPS-693] pixi lock again

sebhmg committed:
[DEVOPS-693] use pixi to run pylint in pre-commit
